src/OpenCLScheduler.cpp

```diff
@@ -254,29 +266,142 @@ template <typename net_t>
+    // critical path).  To do so:
+    //
+    // 1) if we couldn't form a batch after waiting m_waittime ms, it means 
+    // that we hit the critical path and should do scalar evals.
+    // Wait 1ms shorter next time.
+    //
+    // 2) if we picked up a single eval, but were getting additional evals
+    // while that single eval was being processed, it means that we made
+    // the wrong decision.  Wait 2ms longer next time.
+
+    auto pickup_task = [this, gnum] () {
+        std::list<std::shared_ptr<ForwardQueueEntry>> inputs;
+        int count = 0;
+
+        std::unique_lock<std::mutex> lk(m_mutex);
+        while (true) {
+            if (!m_running) return inputs;
+
+            count = static_cast<int>(m_forward_queue.size());
+            if (count >= cfg_batch_size) {
+                count = cfg_batch_size;
                 break;
             }
-            queue_num++;
+
+            bool timeout = !m_cv.wait_for(
+                lk,
+                std::chrono::milliseconds(m_waittime),
+                [this] () {
+                    return !m_running || static_cast<int>(m_forward_queue.size()) >= cfg_batch_size;
+                }
+            );
+
+            if (!m_forward_queue.empty()) {
+                if (timeout && m_single_eval_in_progress.exchange(true) == false) {
+                    // waited long enoough but couldn't form a batch.
+                    // check if there is any other single eval in progress, and if not,
+                    // do one from this thread.
+                    if (m_waittime > 1) {
+                        m_waittime--;
+                    }
+                    count = 1;
+                    break;
+                }
+            }
         }
-        // If this failed, it means we ran out of contexts
-        // which should be more than or equal to the number of threads.
-        assert(ctx != nullptr);
-    }
+        // move 'count' evals from shared queue to local list
+        auto end = begin(m_forward_queue);
+        std::advance(end, count);
+        std::move(begin(m_forward_queue), end, std::back_inserter(inputs));
+        m_forward_queue.erase(begin(m_forward_queue), end);
 
-    m_networks[ctx->net_index]->forward(input, output_pol, output_val,
-                                        ctx->context);
+        return inputs;
+    };
 
-    {
-        LOCK(m_context_pool_mutex, lock);
-        m_context_pool[queue_num].push_back(std::move(ctx));
+    auto batch_input = std::vector<float>();
+    auto batch_output_pol = std::vector<float>();
+    auto batch_output_val = std::vector<float>();
+
+    while (true) {
+        auto inputs = pickup_task();
+        auto count = inputs.size();
+
+        if (!m_running) {
+            return;
+        }
+
+#ifndef NDEBUG
+        batch_stats[static_cast<int>(count) == cfg_batch_size ? 1 : 0]++;
+#endif
+
+        // parepare input for forward() call
+        batch_input.resize(in_size * count);
```

Comment:
If we want to lessen the malloc overhead, these could be thread_locals.

I'm not sure the speed difference will be measurable.

Reply:
Doesn't really need to be thread_local as this local variable will be alive until the thread terminates.  We can avoid resize() but I left it as-is since this isn't something big enough to affect performance.  What probably will happen is that the vector will keep memory large enough to keep the whole batch, while only the (internal) size variable is going to change.